docs/error-type.md

```diff
@@ -0,0 +1,94 @@
+# Solidity `error` type
+
+## Motivation
+
+We want to support typed errors in a syntactically equivalent and convenient way like
+the programmer can define `event` types.
+
+Example:
+
+```solidity
+contract C {
+	event HighestBidIncreased(address bidder, uint amount);
+	error BiddingFailed(string reason);
+
+	function bid() public payable {
+		try attemptBid() {
+			emit HighestBidIncreased(msg.sender, msg.value); // Triggering event
+		} catch BiddingFailed(reason) {
+			// handle error with given reason
+		}
+    }
+
+	function attemptBid() {
+		// assert() should now support a custom secondary parameter to throw custom errors upon
+		failed expectations.
+		assert(middingPossible, BiddingFailed("Bidding impossible"));
```

Comment:
`assert` cannot fail with an error - you need `require` here.

Do we also want to have `throw BiddingFailed("Bidding impossible")`?

Reply:
Or `revert(BiddingFailed("..."))`?

Reply:
That was my plan, to both cases. Yes!

Reply:
Actually, wrt `throw ...` I just realized we have removed the throw keyword in 0.5.0 in favor of revert/require. I'd potentially not re-introduce `throw` right away, because I think that may lead to confusion / disappointment by end-users.

The syntax `require(condition, ErrorType...)` should be also nicely readable (and would only take one line SLoC's as opposed to `if (cond == false) \n throw ErrorType(...);`.

Reply:
Yeah, maybe re-introducing `throw` is not such a good idea. `revert` makes it more obvious what is actually happening.